From 228f363faec6849efa92164e01e5c21dbb75bf32 Mon Sep 17 00:00:00 2001
From: "Richard W.M. Jones" <rjones@redhat.com>
Date: Thu, 6 Oct 2016 22:13:44 +0100
Subject: [PATCH] Fix broken __put_user macro.

---
 arch/riscv/include/asm/uaccess.h | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/arch/riscv/include/asm/uaccess.h b/arch/riscv/include/asm/uaccess.h
index 9f9fba8..eca7b7d 100644
--- a/arch/riscv/include/asm/uaccess.h
+++ b/arch/riscv/include/asm/uaccess.h
@@ -341,20 +341,21 @@ do {								\
 #define __put_user(x, ptr)					\
 ({								\
 	register int __pu_err = 0;				\
+        __typeof__(*(ptr)) __pu_val = (x);			\
 	__typeof__(*(ptr)) __user *__gu_ptr = (ptr);		\
 	__chk_user_ptr(__gu_ptr);				\
 	switch (sizeof(*__gu_ptr)) {				\
 	case 1:							\
-		__put_user_asm("sb", (x), __gu_ptr, __pu_err);	\
+		__put_user_asm("sb", __pu_val, __gu_ptr, __pu_err);	\
 		break;						\
 	case 2:							\
-		__put_user_asm("sh", (x), __gu_ptr, __pu_err);	\
+		__put_user_asm("sh", __pu_val, __gu_ptr, __pu_err);	\
 		break;						\
 	case 4:							\
-		__put_user_asm("sw", (x), __gu_ptr, __pu_err);	\
+		__put_user_asm("sw", __pu_val, __gu_ptr, __pu_err);	\
 		break;						\
 	case 8:							\
-		__put_user_8((x), __gu_ptr, __pu_err);	\
+		__put_user_8(__pu_val, __gu_ptr, __pu_err);	\
 		break;						\
 	default:						\
 		BUILD_BUG();					\
-- 
2.9.3

